### PR TITLE
[tfgan] Add one sided wasserstein gradient penalty

### PR DESCRIPTION
**Pull request by lgeiger**
This PR adds an optional argument to `tf.contrib.gan.losses.wasserstein_gradient_penalty` which allows for a one sided wasserstein gradient penalty.

The paper [On the regularization of Wasserstein GANs](https://arxiv.org/abs/1709.08894) presents theoretical arguments why using a weaker regularization term is preferable over the [two sided gradient penalty](https://arxiv.org/abs/1704.00028) and shows experimental results supporting these arguments.

For reference, [here is the code used in the paper](https://github.com/lukovnikov/improved_wgan_training/blob/master/gan_cifar.py#L144).

/cc @joel-shor @sguada

### CODE REVIEW

**Comment by lgeiger** [comment]
I signed it!

File: tensorflow/contrib/gan/python/losses/python/losses_impl.py [hunk 1]
```diff
@@ -377,9 +383,12 @@ def wasserstein_gradient_penalty(
     # For numerical stability, add epsilon to the sum before taking the square
     # root. Note tf.norm does not add epsilon.
     slopes = math_ops.sqrt(gradient_squares + epsilon)
-    penalties = math_ops.square(slopes / target - 1.0)
+    penalties = slopes / target - 1.0
+    if mode == 'one_sided':
+        penalties = clip_ops.clip_by_value(penalties, 0, np.infty)
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply tensorflow/contrib/gan/python/losses/python/losses_impl.py]: Woops that should've been obvious ;)
👍Fixed

File: tensorflow/contrib/gan/python/losses/python/losses_impl.py [hunk 2]
```diff
@@ -306,6 +306,7 @@ def wasserstein_gradient_penalty(
     discriminator_scope,
     epsilon=1e-10,
     target=1.0,
+    mode='two_sided',
```
[the remark this answers is not in the document]
Comment (joel-shor) [reply tensorflow/contrib/gan/python/losses/python/losses_impl.py]: Instead of adding a string that takes two values, what do you think about adding a boolean argument called something like 'one_sided' that defaults to 'false'?

File: tensorflow/contrib/gan/python/train_test.py [hunk 3]
```diff
@@ -394,6 +395,24 @@ def test_grad_penalty_acgan(self):
   def test_grad_penalty_callable_acgan(self):
     self._test_grad_penalty_helper(create_callable_acgan_model)
 
+  def test_grad_penalty_one_sided_gan(self):
```
Comment (joel-shor) [inline tensorflow/contrib/gan/python/train_test.py]: I strongly applaud your desire to write good tests, but `_test_grad_penalty_helper` only checks that the result is the right kind of object. We don't check a multiple options on other losses, so I don't see these tests as necessary. On the other hand, test-bloat is something to consider. @lgeiger do you have strong feelings on whether to keep or remove these tests? @rmlarsen @sguada do you have thoughts?

**Comment by joel-shor** [comment]
Great addition, @lgeiger ! I'm excited to polish this up and get it submitted.

File: tensorflow/contrib/gan/python/losses/python/losses_impl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply tensorflow/contrib/gan/python/losses/python/losses_impl.py]: I decided to go with string arguments mainly because of two reasons:
1. To be somewhat consistent with the way `tf.layers` handles different modes. E.g. [`data_format` in `tf.layers.Conv2d`](https://www.tensorflow.org/api_docs/python/tf/layers/Conv2D).
2. If down the road there's an additional similar gradient penalty mode this solution would not need a breaking API change since the new mode could just be added to the string set.

That said, I'm happy to switch to a boolean argument if prefered.

File: tensorflow/contrib/gan/python/train_test.py (hunk 3 above)
Comment (lgeiger) [reply tensorflow/contrib/gan/python/train_test.py]: I don't have strong feelings on this. I agree these tests mostly check that the arguments are correctly passed on (I don't know if pylint does this already). Should we keep one of them to prevent trivial regressions?

Comment (rmlarsen) [reply tensorflow/contrib/gan/python/train_test.py]: I wold never argue in favor of removing tests :-)

File: tensorflow/contrib/gan/python/losses/python/losses_impl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmlarsen) [reply tensorflow/contrib/gan/python/losses/python/losses_impl.py]: I don't have a strong opinion about this. I think the string arg is fine, and allows for future additions of new modes (is that likely?). @ebrevdo do you have anything to say?

File: tensorflow/contrib/gan/python/train_test.py (hunk 3 above)
Comment (joel-shor) [reply tensorflow/contrib/gan/python/train_test.py]: Ack. Let's keep them then!

File: tensorflow/contrib/gan/python/losses/python/losses_impl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joel-shor) [reply tensorflow/contrib/gan/python/losses/python/losses_impl.py]: I'm pretty strongly in favor of a boolean, since I think there's a high likelihood that we won't need to bundle more features into this function, and if we do it might not be for a while. I prefer to think of the loss functions as atomic, which new functionality added via new functions (rather than modifying old ones). Your code is an obvious exception since it's such a small variant.

If you remain unconvinced, feel free to leave it. Otherwise, please switch the option here and the one in `train.py` to a boolean.

Thanks!

[the remark this answers is not in the document]
Comment (lgeiger) [reply tensorflow/contrib/gan/python/losses/python/losses_impl.py]: 👍 Switched to a boolean interface.

**Comment by lgeiger** [comment]
Thanks for the fast review. I addressed the comments, this is now ready for testing.

File: tensorflow/contrib/gan/python/losses/python/losses_impl.py [hunk 4]
```diff
@@ -384,7 +381,7 @@ def wasserstein_gradient_penalty(
     # root. Note tf.norm does not add epsilon.
     slopes = math_ops.sqrt(gradient_squares + epsilon)
     penalties = slopes / target - 1.0
-    if mode == 'one_sided':
+    if one_sided is True:
```
Comment (joel-shor) [inline tensorflow/contrib/gan/python/losses/python/losses_impl.py]: `if one_sided is True` -> `if one_sided`?

**Review by joel-shor — CHANGES_REQUESTED** [review]
See comment.

**Review by joel-shor — APPROVED** [review]

File: tensorflow/contrib/gan/python/losses/python/losses_impl.py (hunk 4 above)
Comment (lgeiger) [reply tensorflow/contrib/gan/python/losses/python/losses_impl.py]: Will fix together with lint errors introduced in the last commit. Sorry about that ;)

**Comment by rmlarsen** [comment]
@lgeiger Thanks for the contribution!
